realm/realm-library/src/main/java/io/realm/RealmBaseSchema.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+import java.util.Map;
+import java.util.Set;
+
+import io.realm.internal.ColumnIndices;
+import io.realm.internal.ColumnInfo;
+import io.realm.internal.RealmProxyMediator;
+import io.realm.internal.Table;
+
+
+/**
+ * Class for interacting with the Realm schema using a dynamic API. This makes it possible
+ * to add, delete and change the classes in the Realm.
+ * <p>
+ * All changes must happen inside a write transaction for the particular Realm.
+ *
+ * @see RealmMigration
+ */
+public abstract class RealmBaseSchema {
```

Comment:
The idea about having an abstract super class for the Schema looks like a good idea to me. It means we can split the implementation and gradually move things over/around.

I have some concerns with the naming though. Right now we only expose `RealmSchema` and we should continue to do that in order to not break compatibility.
This means the inheritence hiearchy should looks something like this:

```
public abstract class io.realm.RealmSchema {
	
}

// Our current impl. bikeshed the name
public class StandardRealmSchema extends RealmSchema {
	
}

// The new Object store schema
public class OsRealmSchema extends RealmSchema {
	
}

// Dynamic Realm schema impl if required
public class DynamicRealmSchema extends RealmSchema {
	
}
```

That way we can just return the proper implementation subclass depending on the situation.

Keeping the public interface clean is more important than everything is in the correct package. Right now we have liberal use of package protected methods, which sometimes causes headaches. Classes in `io.realm.internal` are not considered part of the public API so can have whatever modifier makes sense.

Also I'm not sure how this approach solves the backlink field name issue? That seems slightly orthogonal to splitting the code.
A Backlinks field is a kind of ephemeral field, I don't see this proposal addressing this?

One idea would be to have that concept directly in the schema definition. Like a kind of shortcut. That would also enable dynamic realms to take advantage of it, instead of always taking the long route with `linkingObjects(Person.dog).name`:

```
// Just throwing ideas at the wall
//No idea how the concept should be named.
dynamicRealmSchema.setLinkingObjectField("owners", LinkingObjects(dynamicRealmSchema, "Person", "dog"));


// Now a dynamic query can use the same syntax as a typed Realm.
dynamicRealm.where("Dog").equalTo("owners.name", "John")

```